### PR TITLE
feat(server): Enforce TLS >= v1.2

### PR DESCRIPTION
**Pull request by alexec**
Before:
```
(⎈ |k3d-k3s-default:argo)➜  argo-workflows git:(tmp) curl -kv https://localhost:2746/workflows --tls-max 1.1
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 2746 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.1 (OUT), TLS handshake, Client hello (1):
* TLSv1.1 (IN), TLS handshake, Server hello (2):
* TLSv1.1 (IN), TLS handshake, Certificate (11):
* TLSv1.1 (IN), TLS handshake, Server key exchange (12):
* TLSv1.1 (IN), TLS handshake, Server finished (14):
* TLSv1.1 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.1 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.1 (OUT), TLS handshake, Finished (20):
* TLSv1.1 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.1 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.1 / ECDHE-RSA-AES256-SHA
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=localhost; O=ArgoProj
*  start date: Feb 22 19:56:18 2021 GMT
*  expire date: Feb 22 19:56:18 2022 GMT
*  issuer: CN=localhost; O=ArgoProj
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
> GET /workflows HTTP/1.1
> Host: localhost:2746
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
...
```

After:

```
< 
(⎈ |k3d-k3s-default:argo)➜  argo-workflows git:(tmp) curl -kv https://localhost:2746/workflows --tls-max 1.1
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 2746 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.1 (OUT), TLS handshake, Client hello (1):
* TLSv1.1 (IN), TLS alert, protocol version (582):
* error:1400442E:SSL routines:CONNECT_CR_SRVR_HELLO:tlsv1 alert protocol version
* Closing connection 0
curl: (35) error:1400442E:SSL routines:CONNECT_CR_SRVR_HELLO:tlsv1 alert protocol version
```

### CODE REVIEW

File: cmd/argo/commands/server.go [hunk 1]
```diff
@@ -87,8 +88,13 @@ See %s`, help.ArgoSever),
 			if secure {
 				cer, err := tls.LoadX509KeyPair("argo-server.crt", "argo-server.key")
 				errors.CheckError(err)
-				// InsecureSkipVerify will not impact the TLS listener. It is needed for the server to speak to itself for GRPC.
-				tlsConfig = &tls.Config{Certificates: []tls.Certificate{cer}, InsecureSkipVerify: true}
+				tlsMinVersion, err := env.GetInt("TLS_MIN_VERSION", tls.VersionTLS12)
```
Comment (whynowy) [inline cmd/argo/commands/server.go:91]: This ENV needs to be documented somewhere.

File: cmd/argo/commands/server.go [hunk 2]
```diff
@@ -87,8 +88,13 @@ See %s`, help.ArgoSever),
...
+				errors.CheckError(err)
+				tlsConfig = &tls.Config{
+					Certificates:       []tls.Certificate{cer},
+					InsecureSkipVerify: false, // InsecureSkipVerify will not impact the TLS listener. It is needed for the server to speak to itself for GRPC.
```
Comment (whynowy) [inline cmd/argo/commands/server.go:95]: Curious why changing to `false` still works for self-signed CERT.

File: cmd/argo/commands/server.go (hunk 1 above)
Comment (alexec) [reply cmd/argo/commands/server.go:91]: done

File: cmd/argo/commands/server.go (hunk 2 above)
Comment (alexec) [reply cmd/argo/commands/server.go:95]: I think this is because HTTP connections loop back to GRPC, so we need, this

**Comment by whynowy** [comment]
`codegen` keeps failing due to

```
Documented variable TLS_MIN_VERSION in docs/environment-variables.md is not used anywhere
make: *** [codegen] Error 1
Error: Makefile:259: recipe for target 'codegen' failed
```

**Comment by alexec** [comment]
@whynowy fixed. Ready for review.

**Review by whynowy — APPROVED** [review]
LGTM

**Comment by alexec** [comment]
@simster7 I think we should be avoiding back-porting features to v3.0 to now and only do bugs. This is an exception to that mind you.

**Pull request merged**